File: model/src/main/java/no/nav/pensjon/vtp/testmodell/repo/impl/TestscenarioServiceImpl.java
package no.nav.pensjon.vtp.testmodell.repo.impl;

import static java.util.Optional.ofNullable;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Stream;

import org.springframework.stereotype.Component;

import no.nav.pensjon.vtp.testmodell.identer.IdenterIndeks;
import no.nav.pensjon.vtp.testmodell.inntektytelse.InntektYtelseIndeks;
import no.nav.pensjon.vtp.testmodell.load.TestscenarioLoad;
import no.nav.pensjon.vtp.testmodell.organisasjon.OrganisasjonRepository;
import no.nav.pensjon.vtp.testmodell.personopplysning.AdresseIndeks;
import no.nav.pensjon.vtp.testmodell.personopplysning.BrukerModellRepository;
import no.nav.pensjon.vtp.testmodell.personopplysning.PersonIndeks;
import no.nav.pensjon.vtp.testmodell.personopplysning.PersonModell;
import no.nav.pensjon.vtp.testmodell.personopplysning.Personopplysninger;
import no.nav.pensjon.vtp.testmodell.repo.Testscenario;
import no.nav.pensjon.vtp.testmodell.repo.TestscenarioRepository;
import no.nav.pensjon.vtp.testmodell.repo.TestscenarioService;
import no.nav.pensjon.vtp.testmodell.repo.TestscenarioTemplate;

@Component
public class TestscenarioServiceImpl implements TestscenarioService {

    private final TestscenarioFraTemplateMapper mapper;
    private final TestscenarioRepository testscenarioRepository;
    private final PersonIndeks personIndeks;
    private final InntektYtelseIndeks inntektYtelseIndeks;
    private final OrganisasjonRepository organisasjonRepository;
    private final BrukerModellRepository brukerModellRepository;
    private final AdresseIndeks adresseIndeks;
    private final IdenterIndeks identerIndeks;


    public TestscenarioServiceImpl(TestscenarioFraTemplateMapper mapper, TestscenarioRepository testscenarioRepository,
            PersonIndeks personIndeks, InntektYtelseIndeks inntektYtelseIndeks, OrganisasjonRepository organisasjonRepository,
            BrukerModellRepository brukerModellRepository, AdresseIndeks adresseIndeks, IdenterIndeks identerIndeks) {
        this.mapper = mapper;
        this.testscenarioRepository = testscenarioRepository;
        this.personIndeks = personIndeks;
        this.inntektYtelseIndeks = inntektYtelseIndeks;
        this.organisasjonRepository = organisasjonRepository;
        this.brukerModellRepository = brukerModellRepository;
        this.adresseIndeks = adresseIndeks;
        this.identerIndeks = identerIndeks;
    }


    @Override
    public Testscenario opprettTestscenario(TestscenarioTemplate template) {
        return opprettTestscenario(template, Collections.emptyMap());
    }

    @Override
    public Testscenario opprettTestscenario(TestscenarioTemplate template, Map<String, String> variables) {
        return doSave(mapper.lagTestscenario(template, variables));
    }

    @Override
    public Testscenario opprettTestscenarioFraJsonString(String testscenarioJson, Map<String, String> variables) {
        return doSave(mapper.lagTestscenarioFraJsonString(testscenarioJson, variables));
    }

    private Testscenario doSave(TestscenarioLoad testScenario) {
        final String testscenarioId = UUID.randomUUID().toString();
        final Mapper mapper = new Mapper(identerIndeks.getIdenter(testscenarioId), adresseIndeks, testScenario.getVariabelContainer());

        Personopplysninger personopplysninger = mapper.mapFromLoad(testScenario.getPersonopplysninger());

        leggTil(personopplysninger.getSøker());

        ofNullable(personopplysninger.getAnnenPart()).ifPresent(this::leggTil);
        personIndeks.indekserPersonopplysningerByIdent(personopplysninger);

        ofNullable(testScenario.getSøkerInntektYtelse())
                .ifPresent(inntektYtelseModell -> inntektYtelseIndeks.leggTil(personopplysninger.getSøker().getIdent(), inntektYtelseModell));

        ofNullable(personopplysninger.getAnnenPart())
                .ifPresent(annenPart ->
                    ofNullable(testScenario.getAnnenpartInntektYtelse())
                            .ifPresent(inntektYtelseModell -> inntektYtelseIndeks.leggTil(annenPart.getIdent(), inntektYtelseModell))
                );

        organisasjonRepository.saveAll(testScenario.getOrganisasjonModeller());

        return testscenarioRepository.save(mapToTestscenario(testScenario, testscenarioId, personopplysninger));
    }

    private Testscenario mapToTestscenario(final TestscenarioLoad load, final String testscenarioId, final Personopplysninger personopplysningerSave) {
        return new Testscenario(
                load.getTemplateNavn(),
                testscenarioId,
                personopplysningerSave,
                load.getSøkerInntektYtelse(),
                load.getAnnenpartInntektYtelse(),
                load.getOrganisasjonModeller(),
                load.getVariabelContainer().getVars()
        );
    }

    public synchronized void leggTil(PersonModell bruker) {
        if (bruker == null) {
            // quiet escape
            return;
        }

        brukerModellRepository.save(bruker);
    }

    @Override
    public Stream<Testscenario> findAll() {
        return testscenarioRepository.findAll();
    }

    @Override
    public Optional<Testscenario> getTestscenario(String id) {
        return testscenarioRepository.findById(id);
    }

    @Override
    public void slettScenario(String id) {
        testscenarioRepository.delete(id);
    }
}

File: vtp-pensjon-application/src/main/kotlin/no/nav/pensjon/vtp/testmodell/repo/impl/TestscenarioServiceImpl.kt
package no.nav.pensjon.vtp.testmodell.repo.impl

import no.nav.pensjon.vtp.mocks.tp.TjenestepensjonService
import no.nav.pensjon.vtp.testmodell.dkif.DkifRepository
import no.nav.pensjon.vtp.testmodell.dkif.dkifModellMapper
import no.nav.pensjon.vtp.testmodell.identer.IdenterIndeks
import no.nav.pensjon.vtp.testmodell.inntektytelse.InntektYtelseIndeks
import no.nav.pensjon.vtp.testmodell.load.TestscenarioLoad
import no.nav.pensjon.vtp.testmodell.organisasjon.OrganisasjonRepository
import no.nav.pensjon.vtp.testmodell.personopplysning.*
import no.nav.pensjon.vtp.testmodell.repo.*
import no.nav.pensjon.vtp.testmodell.scenario.pensjon.Forhold
import no.nav.pensjon.vtp.testmodell.scenario.pensjon.PensjonTestScenario
import no.nav.pensjon.vtp.testmodell.scenario.pensjon.Tjenestepensjon
import no.nav.pensjon.vtp.testmodell.scenario.pensjon.Ytelse
import no.nav.pensjon.vtp.util.toLocalDate
import org.springframework.stereotype.Component
import java.util.UUID.randomUUID

@Component
class TestscenarioServiceImpl(
    private val mapper: TestscenarioFraTemplateMapper,
    private val testscenarioRepository: TestscenarioRepository,
    private val personIndeks: PersonIndeks,
    private val inntektYtelseIndeks: InntektYtelseIndeks,
    private val organisasjonRepository: OrganisasjonRepository,
    private val personModellRepository: PersonModellRepository,
    private val dkifRepository: DkifRepository,
    private val adresseIndeks: AdresseIndeks,
    private val identerIndeks: IdenterIndeks,
    private val templateRepository: TestscenarioTemplateRepository,
    private val tjenestepensjonService: TjenestepensjonService,
) : TestscenarioService {
    override fun opprettTestscenario(template: TestscenarioTemplate): Testscenario {
        return opprettTestscenario(template, emptyMap())
    }

    override fun opprettTestscenario(
        template: TestscenarioTemplate,
        userSuppliedVariables: Map<String, String>
    ): Testscenario {
        return doSave(mapper.lagTestscenario(template, userSuppliedVariables))
    }

    override fun opprettTestscenarioFraJsonString(
        testscenarioJson: String,
        userSuppliedVariables: Map<String, String>
    ): Testscenario {
        return doSave(mapper.lagTestscenarioFraJsonString(testscenarioJson, userSuppliedVariables))
    }

    private fun doSave(testScenario: TestscenarioLoad): Testscenario {
        val testscenarioId = randomUUID().toString()
        val mapper = Mapper(identerIndeks.getIdenter(testscenarioId), adresseIndeks, testScenario.variabelContainer)
        val personopplysninger = mapper.mapFromLoad(testScenario.personopplysninger)

        leggTil(personopplysninger.søker)
        personopplysninger.annenPart?.let { leggTil(it) }

        personIndeks.indekserPersonopplysningerByIdent(personopplysninger)
        testScenario.søkerInntektYtelse
            ?.let { inntektYtelseIndeks.leggTil(personopplysninger.søker.ident, it) }

        personopplysninger.annenPart
            ?.let { (ident) ->
                testScenario.annenpartInntektYtelse
                    ?.let { inntektYtelseIndeks.leggTil(ident, it) }
            }

        organisasjonRepository.saveAll(testScenario.organisasjonModeller)

        dkifRepository.saveAll(dkifModellMapper(testScenario.dkifModeller, identerIndeks.getIdenter(testscenarioId)))

        return testscenarioRepository.save(mapToTestscenario(testScenario, testscenarioId, personopplysninger))
    }

    private fun mapToTestscenario(
        load: TestscenarioLoad,
        testscenarioId: String,
        personopplysningerSave: Personopplysninger
    ): Testscenario {
        return Testscenario(
            testscenarioId, load.templateNavn,
            personopplysningerSave,
            load.søkerInntektYtelse,
            load.annenpartInntektYtelse,
            load.organisasjonModeller,
            load.dkifModeller,
            load.variabelContainer.getVars()
        )
    }

    fun leggTil(bruker: PersonModell) {
        personModellRepository.save(bruker)
    }

    override fun findAll(): List<Testscenario> {
        return testscenarioRepository.findAll()
    }

    override fun getTestscenario(id: String): Testscenario? {
        return testscenarioRepository.findById(id)
    }

    override fun slettScenario(id: String) {
        testscenarioRepository.deleteById(id)
    }

    override fun opprettPensjonTestScenario(templateId: String, input: PensjonTestScenario): PensjonTestScenario? {
        return templateRepository.finn(templateId)
            ?.let { template ->
                val testscenario = opprettTestscenario(template, emptyMap())
                val person = testscenario.personopplysninger.søker
                input.tjenestepensjon?.let { tjenestepensjon ->
                    tjenestepensjonService.save(person.ident, tjenestepensjon)
                }
                getPensjonTestScenario(testscenario.personopplysninger.søker)
            }
    }

    override fun getPensjonTestScenario(person: PersonModell): PensjonTestScenario {
        val tjenestepensjon: no.nav.pensjon.vtp.mocks.tp.Tjenestepensjon? = tjenestepensjonService.findByPid(person.ident)

        return PensjonTestScenario(
            fnr = person.ident,
            fornavn = person.fornavn,
            etternavn = person.etternavn,
            diskresjonskode = person.diskresjonskode?.name,
            dodsdato = person.dødsdato,
            tjenestepensjon = Tjenestepensjon(
                forhold = tjenestepensjon?.forhold?.map { forhold ->
                    Forhold(
                        tpNr = forhold.tpnr,
                        forholdId = forhold.forholdId!!.toLong(),
                        ytelser = forhold.ytelser.map { ytelse ->
                            Ytelse(
                                type = ytelse.ytelseKode!!,
                                ytelseId = ytelse.ytelseId.toLong(),
                                innmeldtFom = ytelse.innmeldtFom.toLocalDate(),
                                iverksattFom = ytelse.iverksattFom?.toLocalDate(),
                                iverksattTom = ytelse.iverksattTom?.toLocalDate(),
                            )
                        }.toSet()
                    )
                }?.toSet()
                    ?: emptySet()
            )
        )
    }
}
